### PR TITLE
[Wallet] Move key derivation logic from GenerateNewKey to DeriveNewChildKey

### PR DESCRIPTION
**Pull request by pstratem**

### CODE REVIEW

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -683,6 +683,7 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
      * Generate a new key
      */
     CPubKey GenerateNewKey();
+    CKey DeriveNewKey(CKeyMetadata &metadata);
```
Comment (instagibbs) [inline src/wallet/wallet.h]: nit: name is awfully close to GenerateNewKey to me. Perhaps something like "DeriveNewHDKey" or "DeriveNewChildKey".

Comment (pstratem) [inline src/wallet/wallet.h]: Done

**Comment by MarcoFalke** [comment]
Concept ACK

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -156,6 +120,47 @@ CPubKey CWallet::GenerateNewKey()
...
+    // for now we use a fixed keypath scheme of m/0'/0'/k
+    CKey key;                      //master key seed (256bit)
+    CExtKey masterKey;             //hd master key
+    CExtKey accountKey;            //key at m/0'
+    CExtKey externalChainChildKey; //key at m/0'/0'
+    CExtKey childKey;              //key at m/0'/0'/<n>'
+
+    // try to get the master key
+    if (!GetKey(hdChain.masterKeyID, key))
+        throw std::runtime_error("CWallet::GenerateNewKey(): Master key not found");
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: `s/CWallet::GenerateNewKey()/CWallet::DeriveNewChildKey`  :-)

**Comment by jonasschnelli** [comment]
Concept ACK. After having done multiple implementation of a subclasses of `CKeyStore` (`CHDKeyStore`), I think this "new" function should slowly be moved toward `CKeyStore` (or a subclass of it). The whole HD stuff could be kept there to not overpolute `CWallet` (because in general hierarchical key derivation can be wallet neutral).

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (pstratem) [inline src/wallet/wallet.cpp]: Indeed

**Comment by jonasschnelli** [comment]
Needs rebase.

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (sipa) [inline src/wallet/wallet.cpp]: Please fix? :)

Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: No need to fix this. A rebase will solve it, IIRC.

**Comment by sipa** [comment]
Code review ACK, but needs rebase.

**Comment by MarcoFalke** [comment]
@pstratem ^ Trivial(?) rebase needed

**Pull request closed without merging**